ipaclient/install/ipa_epn.py

```diff
@@ -0,0 +1,717 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""This tool prepares then sends email notifications to users
+   whose passwords are expiring in the near future.
+"""
+
+
+from __future__ import absolute_import, print_function
+
+import ast
+import grp
+import json
+import os
+import pwd
+import logging
+import smtplib
+
+from collections import deque
+from datetime import datetime, timedelta
+from email.utils import formataddr, formatdate
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from email.header import Header
+
+from ipaclient.install.client import is_ipa_client_installed
+from ipaplatform.paths import paths
+from ipalib import api, errors
+from ipalib.install import sysrestore
+from ipapython import admintool, ipaldap
+from ipapython.dn import DN
+
+
+EPN_CONF = "/etc/ipa/epn.conf"
+EPN_CONFIG = {
+    "smtp_server": "localhost",
+    "smtp_port": 25,
+    "smtp_user": None,
+    "smtp_password": None,
+    "smtp_timeout": 60,
+    "smtp_security": "none",
+    "smtp_admin": "root@localhost",
+    "notify_ttls": "28,14,7,3,1",
+}
+
+logger = logging.getLogger(__name__)
+
+
+def drop_privileges(new_username="daemon", new_groupname="daemon"):
+    """Drop privileges, defaults to daemon:daemon.
+    """
+    try:
+        if os.getuid() != 0:
+            return
+
+        os.setgroups([])
+        os.setgid(pwd.getpwnam(new_username).pw_uid)
+        os.setuid(grp.getgrnam(new_groupname).gr_gid)
+
+        if os.getuid() == 0:
+            raise Exception()
+
+        logger.debug(
+            "Dropped privileges to user=%s, group=%s",
+            new_username,
+            new_groupname,
+        )
+
+    except Exception as e:
+        logger.error(
+            "Failed to drop privileges to %s, %s: %s",
+            new_username,
+            new_groupname,
+            e,
+        )
+
+
+class EPNUserList:
+    """Maintains a list of users whose passwords are expiring.
+       Provides add(), check(), pop(), and json_print().
+       From the outside, the list is considered always sorted:
+       * displaying the list results in a sorted JSON representation thereof
+       * pop() returns the "most urgent" item from the list.
+       Internal implementation notes:
+       * Uses a deque instead of a list for efficiency reasons
+       * all add()-style methods MUST set _sorted to False.
+       * all print() and pop-like methods MUST call _sort() first.
+    """
+
+    def __init__(self):
+        self._sorted = False
+        self._expiring_password_user_dq = deque()
+
+    def __bool__(self):
+        """If it quacks like a container...
+        """
+        return bool(self._expiring_password_user_dq)
+
+    def __len__(self):
+        """Return len(self)."""
+        return len(self._expiring_password_user_dq)
+
+    def add(self, entry):
+        """Parses and appends an LDAP user entry with the uid, cn,
```

Comment:
typo: "an LDAP" -> "a LDAP"

Reply:
I've double-checked and the correct form is "an LDAP". It sounds better to me, but that's all I can say, grammar-wise!